graph.js

```
@@ -1,12 +1,194 @@
+  }
+
+  if (app.node === undefined) {
+    console.log('  Graph node:    No node defined with -g/--node')
+  } else {
+    console.log(`  Graph node:    ${app.node}`)
+  }
+
+  if (app.ipfs === undefined) {
+    console.log('  IPFS:          No node defined with -i/--ipfs')
+  } else {
+    console.log(`  IPFS:          ${app.ipfs}`)
+  }
+
+  console.log('')
+})
+
+app
+  .command('codegen <subgraph-manifest>')
+  .description('Generates TypeScript types for a subgraph')
+  .action(subgraphManifest => {
+    let generator = new TypeGenerator({
+      subgraphManifest,
+      outputDir: app.outputDir,
+      verbosity: app.verbosity,
+    })
+
+    // Watch working directory for file updates or additions, trigger
+    // type generation (if watch argument specified)
+    if (app.watch) {
+      generator.watchAndGenerateTypes()
+    } else {
+      generator.generateTypes()
+    }
+  })
+
+app
+  .command('build <subgraph-manifest>')
+  .description('Compiles a subgraph and uploads it to IPFS')
+  .action(subgraphManifest => {
+    let compiler = createCompiler(app, subgraphManifest)
+
+    // Watch subgraph files for changes or additions, trigger
+    // compile (if watch argument specified)
+    if (app.watch) {
+      compiler.watchAndCompile()
+    } else {
+      compiler.compile()
+    }
+  })
+
+app
+  .command('deploy <subgraph-manifest>')
+  .description('Deploys the subgraph to a graph node')
+  .action(subgraphManifest => {
+    if (app.subgraphName == undefined) {
+      console.error('Error: No subgraph name specified with -n/--subgraph-name')
+      console.error('--')
+      app.help()
+    }
+
+    if (app.node == undefined) {
+      console.error('Error: No Graph node specified with -g/--node')
+      console.error('--')
+      app.help()
+    }
+
+    if (app.ipfs == undefined) {
+      console.error('Error: No IPFS node specified with -i/--ipfs')
+      console.error('--')
+      app.help()
+    }
```

Can we change this so that it report errors for all missing parameters?

Do we really want to print the whole help when there is an error in an argument? The actual error gets lost in the help text. I'd rather print "For more information try --help".

---

Yep, I'll update.